File: SexyBiscuit.Engine/Gameplay/GameInstance.cs
using SexyBiscuit.Engine.Core;

namespace SexyBiscuit.Engine.Gameplay;

/// <summary>
/// The one object that outlives every scene — the natural home for save data, the active
/// player profile, matchmaking state and anything else that must survive a level change.
/// Modelled on Unreal's <c>UGameInstance</c>.
/// </summary>
/// <remarks>
/// Exactly one game instance exists per process. <see cref="SBEngine"/> creates a plain
/// <see cref="GameInstance"/> at startup unless the game supplies its own subclass through
/// <see cref="EngineConfig.GameInstanceFactory"/>.
/// </remarks>
/// <example>
/// <code>
/// public class MyGameInstance : GameInstance
/// {
///     public string PlayerName = "Player";
///     public override void OnStart() =&gt; Console.WriteLine("Session begins");
/// }
///
/// SBEngine.Run(new EngineConfig { GameInstanceFactory = () =&gt; new MyGameInstance() });
/// </code>
/// </example>
public class GameInstance
{
    /// <summary>The active game instance. Never null once the engine has initialised.</summary>
    public static GameInstance Current { get; internal set; } = null!;

    /// <summary>Subsystems scoped to the whole session.</summary>
    public SubsystemCollection Subsystems { get; } = new();

    /// <summary>Raised after <see cref="OnStart"/> completes.</summary>
    public SBEvent Started { get; } = new();

    /// <summary>Raised just before <see cref="OnShutdown"/> runs.</summary>
    public SBEvent ShuttingDown { get; } = new();

    /// <summary>Raised whenever a scene finishes loading, with the new scene.</summary>
    public SBEvent<Core.Scene> WorldChanged { get; } = new();

    internal void InternalInit()
    {
        Current = this;
        Subsystems.Configure = s =>
        {
            if (s is GameInstanceSubsystem gis) gis.GameInstance = this;
        };
    }

    /// <summary>Shorthand for <c>Subsystems.Get&lt;T&gt;()</c>.</summary>
    public T? GetSubsystem<T>() where T : GameInstanceSubsystem, new() => Subsystems.Get<T>();

    /// <summary>Called once when the engine has finished initialising, before the first scene loads.</summary>
    public virtual void OnStart() { }

    /// <summary>Called once as the engine shuts down. Flush saves here.</summary>
    public virtual void OnShutdown() { }

    /// <summary>Called every frame with the scaled delta, before scenes tick.</summary>
    public virtual void Tick(float dt) { }

    internal void InternalStart()
    {
        OnStart();
        Started.Broadcast();
    }

    internal void InternalTick(float dt)
    {
        Tick(dt);
        Subsystems.Tick(dt);
    }

    internal void InternalShutdown()
    {
        ShuttingDown.Broadcast();
        OnShutdown();
        Subsystems.Shutdown();
    }
}

File: html5/src/gameplay/GameInstance.js
// -----------------------------------------------------------------------------
// GameInstance — state that outlives any one scene.
// -----------------------------------------------------------------------------

import { SBEvent } from '../core/SBEvent.js';

/** Base for something that lives alongside the game instance. */
export class Subsystem {
    constructor() {
        this.isInitialized = false;
    }

    /** Return false to skip creating this subsystem. */
    shouldCreate() { return true; }

    initialize() {}
    deinitialize() {}
    tick(dt) {}
}

/** State that survives scene changes: save data, settings, the session. */
export class GameInstance {
    /** The instance the engine host created. */
    static current = null;

    constructor() {
        /** @type {Map<Function, Subsystem>} */
        this.subsystems = new Map();

        this.started = new SBEvent();
        this.shuttingDown = new SBEvent();
        this.worldChanged = new SBEvent();
    }

    /** Adds a subsystem, or returns the existing one. */
    getSubsystem(Ctor) {
        const existing = this.subsystems.get(Ctor);
        if (existing) return existing;

        const subsystem = new Ctor();
        if (!subsystem.shouldCreate()) return null;

        subsystem.initialize();
        subsystem.isInitialized = true;
        this.subsystems.set(Ctor, subsystem);
        return subsystem;
    }

    internalInit() { GameInstance.current = this; }

    internalStart() {
        this.onStart();
        this.started.broadcast();
    }

    internalTick(dt) {
        this.tick(dt);
        for (const subsystem of this.subsystems.values()) {
            if (subsystem.isInitialized) subsystem.tick(dt);
        }
    }

    internalShutdown() {
        this.shuttingDown.broadcast();
        this.onShutdown();
        for (const subsystem of this.subsystems.values()) subsystem.deinitialize();
        this.subsystems.clear();
        if (GameInstance.current === this) GameInstance.current = null;
    }

    // ---- Hooks ---------------------------------------------------------------

    onStart() {}
    onShutdown() {}
    tick(dt) {}
}
